libcnb-test/src/test_runner.rs

```diff
@@ -134,21 +134,29 @@ impl TestRunner {
     ) {
         let config = config.borrow();
 
-        let app_dir = if config.app_dir.is_relative() {
-            env::var("CARGO_MANIFEST_DIR")
-                .map(PathBuf::from)
-                .expect("Could not determine Cargo manifest directory")
-                .join(&config.app_dir)
-        } else {
-            config.app_dir.clone()
-        };
+        let app_dir = {
+            let normalized_app_dir_path = if config.app_dir.is_relative() {
+                env::var("CARGO_MANIFEST_DIR")
+                    .map(PathBuf::from)
+                    .expect("Could not determine Cargo manifest directory")
+                    .join(&config.app_dir)
+            } else {
+                config.app_dir.clone()
+            };
 
-        let temp_app_dir =
-            app::copy_app(&app_dir).expect("Could not copy app to temporary location");
+            // Copy the app to a temporary directory if an app_dir_preprocessor is specified and run the
+            // preprocessor. Skip app copying if no changes to the app will be made.
+            if let Some(app_dir_preprocessor) = &config.app_dir_preprocessor {
+                let app_dir = app::copy_app(&normalized_app_dir_path)
+                    .expect("Could not copy app to temporary location");
 
-        if let Some(app_dir_preprocessor) = &config.app_dir_preprocessor {
-            (app_dir_preprocessor)(PathBuf::from(temp_app_dir.path()));
-        }
+                (app_dir_preprocessor)(app_dir.as_path().to_owned());
+
+                app_dir
```

Comment:
(Optional) Would renaming `app_dir` (just inside this block) to `temporary_app_dir` make it clearer what's different about what's being returned in this codepath? 

It's just at the moment, if it were not for the `.expect()` message, it wouldn't be clear to me that this is a temporary directory.